Sources/Plasma/CoreLib/hsMatrixMath.h

```diff
@@ -51,13 +51,27 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
 #import <Accelerate/Accelerate.h>
 #endif
 
+#ifdef HS_BUILD_FOR_WIN32
+#include   <DirectXMath.h>
+#endif
+
 static inline hsMatrix44 IMatrixMul34(const hsMatrix44& lhs, const hsMatrix44& rhs)
 {
     hsMatrix44 ret;
     ret.NotIdentity();
-    
-#ifdef HS_BUILD_FOR_APPLE
+
+#if defined(HS_BUILD_FOR_APPLE)
     vDSP_mmul((const float*)lhs.fMap, 1, (const float*)rhs.fMap, 1, (float*)&(ret.fMap), 1, 3, 4, 4);
+    ret.fMap[3][0] = ret.fMap[3][1] = ret.fMap[3][2] = 0.f;
+    ret.fMap[3][3] = 1.f;
+#elif defined(HS_BUILD_FOR_WIN32)
+    // Trying to interpret an hsMatrix44's fMap as an XMFLOAT3X4 doesn't work, sadly.
+    // This is still a win for performance over the auto-vectorized FPU version below,
+    // especially once the manual initialization is removed.
+    DirectX::XMMATRIX m1 = DirectX::XMLoadFloat4x4A((DirectX::XMFLOAT4X4A*)lhs.fMap);
+    DirectX::XMMATRIX m2 = DirectX::XMLoadFloat4x4A((DirectX::XMFLOAT4X4A*)rhs.fMap);
+    DirectX::XMMATRIX result = m1 * m2;
+    DirectX::XMStoreFloat4x4A((DirectX::XMFLOAT4X4A*)ret.fMap, result);
```

Comment:
Isn't this missing the bit where it sets `fMap[3]` to `{0, 0, 0, 1}` at the end?

Or, if that is not in fact needed, is there a reason not to treat all of these like 4×4 multiplications?

Reply:
The reason why these are not 4x4 matrices is because it's a special case optimization for when the entire matrix is rarely an identity matrix and only the first three rows are used in the result. So it is an optimization to ignore the 4th row in the math and avoid the branching in the main multiplication code path. Unfortunately, `DirectXMath` is using its own float array struct that isn't laid out quite like ours in the 3x4 case, so we must do a full-fat 4x4 multiplication here. The full fat multiplication gets the proper `0, 0, 0, 1` value from the input matrices, so the result's `fMap[3]` will be `0, 0, 0, 1` automatically. The other code paths must do this initialization manually because they do not perform the full 4x4 multiplication, and `hsMatrix44` does not have a default initializer (probably for performance reasons).

In my testing `DirectXMath`'s full 4x4 multiplication is faster than the FPU 3x4 multiplication + `fMap[3]` initialization.